File: 542.01-matrix.go
/*
 * @lc app=leetcode id=542 lang=golang
 *
 * [542] 01 Matrix
 */

// @lc code=start
func updateMatrix(mat [][]int) [][]int {
	m, n := len(mat), len(mat[0])
    res := make([][]int, m)
    for i := range res {
        res[i] = make([]int, n)
        for j := range res[i] {
            res[i][j] = math.MaxInt32
        }
    }
    
    q := make([][]int, 0)
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            if mat[i][j] == 0 {
                res[i][j] = 0
                q = append(q, []int{i, j})
            }
        }
    }
    q = append(q, nil)
    
    shifts := [][]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
    visited := make([][]bool, m)
    for i := range visited {
        visited[i] = make([]bool, n)
    }
    
    dist := 1
    for !(len(q) == 1 && len(q[0]) == 0) {
        if len(q[0]) == 0 {
            dist++
            q = q[1:]
            q = append(q, nil)
        }
        
        zero := q[0]
        q = q[1:]
        for _, shift := range shifts {
            x, y := zero[0]+shift[0], zero[1]+shift[1]
            if x < 0 || x >= m || y < 0 || y >= n {
                continue
            }
            if visited[x][y] {
                continue
            }
            if dist < res[x][y] {
                res[x][y] = dist
                q = append(q, []int{x, y})
            }
        }
        visited[zero[0]][zero[1]] = true
    }
    
    return res
}
// @lc code=end

File: 542.01-matrix.js
/*
 * @lc app=leetcode id=542 lang=javascript
 *
 * [542] 01 Matrix
 */

// @lc code=start
/**
 * @param {number[][]} mat
 * @return {number[][]}
 */
var updateMatrix = function (mat) {
    const res = Array(mat.length).fill([]).map(() => Array(mat[0].length).fill(1e5))
    const queue = [];
    for (let i = 0; i < mat.length; i++) {
        for (let j = 0; j < mat[0].length; j++) {
            if (mat[i][j] == 0) {
                res[i][j] = 0;
                queue.push([i, j]);
            }
        }
    }
    queue.push(null)

    const shifts = [[0, 1], [0, -1], [1, 0], [-1, 0]]
    const visited = Array(mat.length).fill([]).map(() => Array(mat[0].length).fill(false))
    dist = 1
    while (!(queue.length == 1 && queue[0] == null)) {
        if (queue[0] == null) {
            queue.shift()
            queue.push(null)
            dist++;
        }

        const zero = queue.shift()
        for (const shift of shifts) {
            const x = zero[0] + shift[0]
            const y = zero[1] + shift[1]
            if (x < 0 || x >= mat.length || y < 0 || y >= mat[0].length) {
                continue
            }
            if (visited[x][y]) {
                continue
            }
            if (dist < res[x][y]) {
                res[x][y] = dist
                queue.push([x, y])
            }
        }

        visited[zero[0]][zero[1]] = true
    }

    return res;
};
// @lc code=end

